jpw1991/chebs-necromancy
# minion item drops & wait

# Minion Item Drops & Wait

If you don't want to/can't build your own version, you can take this pre-release version [here](https://github.com/jpw1991/chebs-necromancy/releases/tag/1.6.5). Be sure to test this in a new world and not your primary one in case of bugs.

## Wait

Wait now works properly -> minions no longer wander off

### Testing

- Tell minions to wait
- Ensure that they do actually wait and don't wander around
- Ensure they still function properly eg. being able to attack and what not
- Ensure that they remember where they're waiting after logging off & coming back

### Mechanism

- Reduced random wander distance on AI to 0
- When minions get told to wait, they write the position into the ZDO and spawn an invisible object there and follow it
- When session is restored, so is the invisible object and they're told to follow it again
- Telling them to follow will destroy the invisible object and make them follow the player

## Minion Items Drops

Minions can be configured to drop items

### Mechanism

- Upon minion creation, the CharacterDrop component is added and resources are recorded in the ZDO as a delimited string eg. `BoneFragments:3,Iron:1,SurtlingCore:1`
- When the session ends, the CharacterDrop component is lost
- Upon session being restored, the ZDO is read and parsed and the CharacterDrop is added again with the recorded items

### Four new config entries (one for Skeletons, one for Draugr)

Sample:

```
[SkeletonMinion (Server Synced)]

## Whether a minion refunds anything when it dies.
# Setting type: DropType
# Default value: JustResources
# Acceptable values: Nothing, JustResources, Everything
DropOnDeath = JustResources

## If set to true, dropped items will be packed into a cargo crate. This means they won't sink in water, which is useful for more valuable drops like Surtling Cores and metal ingots.
# Setting type: Boolean
# Default value: true
PackDroppedItemsIntoCargoCrate = true
```
### Testing

- Skeleton
  - `DropOnDeath = Nothing`
    - Current/standard behavior should be exhibited.
  - `DropOnDeath = JustResources`
    - If you make an iron skeleton and it costs 1 bar, you get that back when it dies.
  - `DropOnDeath = Everything`
    - Everything used to create a skeleton should be dropped on death: bones, ingots, surtling cores.
  - `PackDroppedItemsIntoCargoCrate = true`
    - The items should get packed into a buoyant box
  - `PackDroppedItemsIntoCargoCrate = false`
    - The items should get dumped on the floor
- Draugr
  - The equivalent tests for Draugr
